hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/ExchangeRateInfoImpl.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.fees;
+
+import static java.util.Objects.requireNonNull;
+
+import com.hedera.hapi.node.transaction.ExchangeRate;
+import com.hedera.hapi.node.transaction.ExchangeRateSet;
+import com.hedera.node.app.spi.fees.ExchangeRateInfo;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.time.Instant;
+
+/**
+ * Default implementation of {@link ExchangeRateInfo}.
+ */
+public record ExchangeRateInfoImpl(@NonNull ExchangeRateSet exchangeRateSet) implements ExchangeRateInfo {
+
+    public ExchangeRateInfoImpl {
+        requireNonNull(exchangeRateSet, "exchangeRateSet must not be null");
+        requireNonNull(exchangeRateSet.currentRate(), "exchangeRateSet.currentRate() must not be null");
+        requireNonNull(
+                exchangeRateSet.currentRateOrThrow().expirationTime(),
+                "exchangeRateSet.currentRate().expirationTime() must not be null");
+        requireNonNull(exchangeRateSet.nextRate(), "exchangeRateSet.nextRate() must not be null");
+    }
+
+    @NonNull
+    @Override
+    public ExchangeRateSet exchangeRates() {
+        return exchangeRateSet;
+    }
+
+    @NonNull
+    @Override
+    public ExchangeRate activeRate(@NonNull Instant consensusTime) {
+        return consensusTime.getEpochSecond()
+                        > exchangeRateSet
+                                .currentRateOrThrow()
+                                .expirationTimeOrThrow()
+                                .seconds()
+                ? exchangeRateSet.nextRateOrThrow()
+                : exchangeRateSet.currentRateOrThrow();
```

Comment:
If this method is called each TX, maybe worth caching the expiration time of the current rate, and next rate, and current rate. If called once per round, then this is fine.

Reply:
I have changed it. It is a class now that caches the values. I am still wondering, though, if optimizations like this will have any impact at all. There are unnecessary null checks, but branch prediction will always get them right.

All I really wanted to do was dereference the properties of records. Unfortunately, we are forced to use ...OrThrow() or live with many warnings in the IDE. Both are not optional. We should reconsider if we can define non-null fields in PBJ where the `null` check is done only once during instantiation.